SNLS_NLSPowell.h

```diff
@@ -0,0 +1,729 @@
+#pragma once
+
+#include "SNLS_base.h"
+#include "SNLS_qr_solve.h"
+#include "SNLS_linalg.h"
+
+#include <stdlib.h>
+#include <iostream>
+#ifdef __cuda_host_only__
+#include <string>
+#include <sstream>
+#include <iomanip>
+#endif
+
+namespace snls {
+
+/** Helper templates to ensure compliant CRJ implementations */
+template<typename CRJ, typename = void>
+struct nls_has_valid_computeRJ : std::false_type { static constexpr bool value = false;};
+
+template<typename CRJ>
+struct nls_has_valid_computeRJ <
+   CRJ,typename std::enable_if<
+       std::is_same<
+           decltype(std::declval<CRJ>().computeRJ(std::declval<double* const>(), std::declval<double* const>(),std::declval<const double *>())),
+           bool  
+       >::value
+       ,
+       void
+   >::type
+>: std::true_type { static constexpr bool value = true;};
+
+template<typename CRJ, typename = void>
+struct nls_has_ndim : std::false_type { static constexpr bool value = false;};
+
+template<typename CRJ>
+struct nls_has_ndim <
+   CRJ,typename std::enable_if<
+       std::is_same<
+           decltype(CRJ::nDimSys),
+           const int  
+       >::value
+       ,
+       void
+   >::type
+>: std::true_type { static constexpr bool value = true;};
+
+// Hybrid Jacobian method from MINPACK which is a modified Powell method
```

Comment:
So, the algorithm itself would be fine, but we might need to see how things are originally defined in the original paper/conference proceedings in regards to the exit conditions of the algorithm. I've ordered the conference book as the chapter is impossible to find online..., so I'll know the answer to that sometime next week. I did also find the following MINPACK document that actually goes over all the various exit conditions and algorithms as well: http://cds.cern.ch/record/126569/files/CM-P00068642.pdf. I have done somethings differently, but the exit conditions are largely the same as what's defined by minpack...

Outside that it appears that MINPACK (http://www.netlib.org/minpack/disclaimer) has essentially a BSD type license, so it might be fairly simple to include this in the header but as per your later comment we should probably go through another release process for this just to make sure things are good.

Reply:
With the last bit of refactoring, I would say we are quite a bit different than that we should no longer have to worry about any copyright issues. I've also removed all these sorts of references outside the one at the top that mentions that we used MINPACK, the original Powell paper, and our own implementation as inspiration for this solver.